### ISSUE TITLE
runtime: slice bounds out of range in issue9321

### ISSUE THREAD

**Issue by RLH** [golang/go#9714]
While continually running all.bash overnight I noticed that the mumble/go/test/fixedbugs/issue9321.go test failed after about 12 hours. I isolated the test and then started running the test using a simple loop
while ./issue9321; do true; done 
This would reliably fail within about 10 minutes. This is on tip on an x64 Ubuntu 12 core Linux Google desktop machine.

```
panic: runtime error: slice bounds out of range

goroutine 9 [running]:
bytes.(*Buffer).grow(0xc208010070, 0x6c6, 0xc208166a20)
	/usr/local/google/home/rlh/work/go/src/bytes/buffer.go:105 +0x189
bytes.(*Buffer).Write(0xc208010070, 0xc2085b4000, 0x6c6, 0x100000, 0x6c6, 0x0, 0x0)
	/usr/local/google/home/rlh/work/go/src/bytes/buffer.go:127 +0x52
runtime/pprof.writeGoroutineStacks(0x7f06369fbf88, 0xc208010070, 0x0, 0x0)
	/usr/local/google/home/rlh/work/go/src/runtime/pprof/pprof.go:522 +0xf0
runtime/pprof.writeGoroutine(0x7f06369fbf88, 0xc208010070, 0x2, 0x0, 0x0)
	/usr/local/google/home/rlh/work/go/src/runtime/pprof/pprof.go:500 +0x4f
runtime/pprof.(*Profile).WriteTo(0x5aeb60, 0x7f06369fbf88, 0xc208010070, 0x2, 0x0, 0x0)
	/usr/local/google/home/rlh/work/go/src/runtime/pprof/pprof.go:229 +0xd5
main.func·001()
	/usr/local/google/home/rlh/work/go/test/fixedbugs/issue9321.go:22 +0xce
created by main.test
	/usr/local/google/home/rlh/work/go/test/fixedbugs/issue9321.go:28 +0xbc

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc20800a2e0)
	/usr/local/google/home/rlh/work/go/src/runtime/sema.go:43 +0x2d
sync.(*WaitGroup).Wait(0xc208058020)
	/usr/local/google/home/rlh/work/go/src/sync/waitgroup.go:132 +0x169
main.test()
	/usr/local/google/home/rlh/work/go/test/fixedbugs/issue9321.go:29 +0xca
main.main()
	/usr/local/google/home/rlh/work/go/test/fixedbugs/issue9321.go:35 +0x39

goroutine 8 [runnable]:
runtime/pprof.writeGoroutineStacks(0x7f06369fbf88, 0xc2080400e0, 0x0, 0x0)
	/usr/local/google/home/rlh/work/go/src/runtime/pprof/pprof.go:509 +0x4e
runtime/pprof.writeGoroutine(0x7f06369fbf88, 0xc2080400e0, 0x2, 0x0, 0x0)
	/usr/local/google/home/rlh/work/go/src/runtime/pprof/pprof.go:500 +0x4f
runtime/pprof.(*Profile).WriteTo(0x5aeb60, 0x7f06369fbf88, 0xc2080400e0, 0x2, 0x0, 0x0)
	/usr/local/google/home/rlh/work/go/src/runtime/pprof/pprof.go:229 +0xd5
main.func·001()
	/usr/local/google/home/rlh/work/go/test/fixedbugs/issue9321.go:22 +0xce
created by main.test
	/usr/local/google/home/rlh/work/go/test/fixedbugs/issue9321.go:27 +0xa5
```

**Comment by bradfitz** [comment]
What's the race detector say?

**Comment by bradfitz** [comment]
Just tried it... the program looks race-free, and the race detector agrees.

**Comment by randall77** [comment]
I ran all weekend and saw no failures.
Something must be different about our setup.  GOMAXPROCS?  Maybe it wasn't tip?

**Comment by aclements** [comment]
I've run through 45,000+ iterations of this test on what I believe is the same hardware Rick used and haven't had any failures. I suspect whatever this was has been fixed in the intervening months. I also went through the cases in Buffer.grow and there does not appear to be any way to get an out-of-bounds in correctly executing code, suggesting this was a bug in the compiler or runtime.

@RLH, since you were able to reproduce it easily before, care to give it another quick test and reopen if it's still happening?
